### ISSUE TITLE
Cache during indexing?

### ISSUE THREAD

**Issue by hobofan** [weaviate/weaviate#2618]
Hey there!

Occasionally when indexing documents, we have a large number of duplicates for values to be vectorized. (E.g. 90% of values are one of 5 values because they come from a fixed system, and the remaining 10% are more "organic" user-entered data).

From what I could tell it looks like for the duplicate values, redundant requests are being sent to the vectorizer that have identical requests and responses. That to me looks like something that could be cached pretty well, or be solved by deduplicating identical values inside one batch indexing request.

My first instinct on how to tackle that would be to put e.g. a Varnish cache between weaviate and the vectorizer container. Alternatively, I think a caching middleware in the vectorizer container might also be a rather lightweight approach that wouldn't complicate the weaviate core codebase.

I was wondering what your thoughts on this are, and whether maybe someone has already made some experience with either approach?

P.S.: Thanks a lot for building Weaviate! Picking it up and integrating it for some first demos has been a breeze for us!

**Comment by etiennedi** [comment]
> P.S.: Thanks a lot for building Weaviate! Picking it up and integrating it for some first demos has been a breeze for us!

Love hearing that, thank you!

Before diving deeper into the feasibility of a cache, let me get a better understanding of this part:

> Occasionally when indexing documents, we have a large number of duplicates for values to be vectorized. (E.g. 90% of values are one of 5 values because they come from a fixed system, and the remaining 10% are more "organic" user-entered data).

Can you elaborate on what level that 90/10 split occurs? 
- If you have a single object where 90% of the props are identical and 10% are not, then the vectorization request would not be cachable, because you need all of it the build the vector
- If 90% are identical objects and 10% are diverse objects, then you introduce another problem: ANN indexes can't deal well with identical objects. (In short: HNSW is a proximity graph that relies on the variance of distances for its edges. If you have identical vectors, you end up with local clusters where all edges are of length 0 and you can't move efficiently in the graph anymore). In this case, it might be better to have the unique objects imported just once each and use a cross-reference to link to them (a bit like a normalized pattern in a SQL db).

**Comment by hobofan** [comment]
> If 90% are identical objects and 10% are diverse objects, then you introduce another problem: ANN indexes can't deal well with identical objects.

Yes, it's that case. As we have quite a flexible schema in our product, and the way we've integrated it right now basically would allow a customer to trigger indexing and similarity search in a "Give me similar entities of X based on property Y" was, that situation doesn't really seem avoidable...

I didn't realize that that degrades the index. I guess that might also explain why for indexes with largely identical objects there is a slowdown in indexing? (e.g. when indexing some objects with a 95% identical text property with ~10 characters each, indexing took 2x as long as indexing the same object with a highly diverse property of ~50 characters each)

> In this case, it might be better to have the unique objects imported just once each and use a cross-reference to link to them (a bit like a normalized pattern in a SQL db).

Yeah, I guess that would be possible (although a bit more cumbersome and what I hoped to avoid 😅 ). Thanks for the advice! :)
